README.md

[![GitHub release (latest by date)](https://img.shields.io/github/v/release/xpack-dev-tools/xbb-bootstrap-xpack)](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/releases)
[![npm (scoped)](https://img.shields.io/npm/v/@xpack-dev-tools/xbb-bootstrap.svg)](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap/)

# The xPack XBB Bootstrap

A standalone cross-platform (macOS/Linux) **XBB Bootstrap**
binary distribution, intended for reproducible builds.

It includes most of the tools available in the XBB Docker v3.4 images,
and is intended to be used to bootstrap the creation of separate
packages with these tools.

In addition to the the binary archives and the package meta data,
this project also includes the build scripts.

## Early deprecation notice

This package is not recommended for new designs. It will be used
as a direct replacement for the Docker images, to build all existing
binary xPacks.

Once the new build scripts are functional, the plan is to move
components out as separate packages, to the point when there will be
no components left, when this package will be retired.

## Overview

This open source project is hosted on GitHub as
[`xpack-dev-tools/xbb-bootstrap-xpack`](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack)
and provides the platform specific binaries for the
[xPack XBB Bootstrap](https://xpack.github.io/xbb-bootstrap/).

The binaries can be installed automatically as **binary xPacks** or manually as
**portable archives**.

## Release schedule

There are no scheduled releases.

## User info

This section is intended as a shortcut for those who plan
to use the XBB Bootstrap binaries. For full details please read the
[xPack XBB](https://xpack.github.io/xbb/) pages.

### Easy install

The easiest way to install XBB Bootstrap is using the **binary xPack**, available as
[`@xpack-dev-tools/xbb-bootstrap`](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap)
from the [`npmjs.com`](https://www.npmjs.com) registry.

#### Prerequisites

The only requirement is a recent
`xpm`, which is a portable
[Node.js](https://nodejs.org) command line application. To install it,
follow the instructions from the
[xpm](https://xpack.github.io/xpm/install/) page.

#### Install

With the `xpm` tool available, installing
the latest version of the package and adding it as
a dependency for a project is quite easy:

```sh
cd my-project
xpm init # Only at first use.

xpm install @xpack-dev-tools/xbb-bootstrap@latest

ls -l xpacks/.bin
```

This command will:

- install the latest available version,
into the central xPacks store, if not already there
- add symbolic links to the central store
into the local `xpacks/.bin` folder.

The central xPacks store is a platform dependent
folder; check the output of the `xpm` command for the actual
folder used on your platform).
This location is configurable via the environment variable
`XPACKS_STORE_FOLDER`; for more details please check the
[xpm folders](https://xpack.github.io/xpm/folders/) page.

For xPacks aware tools, like the **Eclipse Embedded C/C++ plug-ins**,
it is also possible to install XBB Bootstrap globally, in the user home folder:

```sh
xpm install --global @xpack-dev-tools/xbb-bootstrap@latest
```

#### Uninstall

To remove the links from the current project:

```sh
cd my-project

xpm uninstall @xpack-dev-tools/xbb-bootstrap
```

To completely remove the package from the global store:

```sh
xpm uninstall --global @xpack-dev-tools/xbb-bootstrap
```

### Manual install

For all platforms, the **xPack XBB Bootstrap**
binaries are released as portable
archives that can be installed in any location.

The archives can be downloaded from the
GitHub [Releases](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/releases/)
page.

For more details please read the
[Install](https://xpack.github.io/xbb-bootstrap/install/) page.

### Versioning

The version strings used by the XBB project are semver strings
like `4.0.0`.

## Maintainer info

- [How to build](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/blob/xpack/README-BUILD.md)
- [How to make new releases](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/blob/xpack/README-RELEASE.md)
- [Developer info](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/blob/xpack/README-DEVELOP.md)

## Support

The quick answer is to use the GitHub
[Discussions](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/discussions/).

For more details please read the
[Support](https://xpack.github.io/xbb-bootstrap/support/) page.

## License

The original content is released under the
[MIT License](https://opensource.org/licenses/MIT), with all rights
reserved to [Liviu Ionescu](https://github.com/ilg-ul/).

The binary distributions include several open-source components; the
corresponding licenses are available in the installed
`distro-info/licenses` folder.

## Download analytics

- GitHub [`xpack-dev-tools/xbb-bootstrap-xpack`](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/) repo
  - latest xPack release
[![Github All Releases](https://img.shields.io/github/downloads/xpack-dev-tools/xbb-bootstrap-xpack/latest/total.svg)](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/releases/)
  - all xPack releases [![Github All Releases](https://img.shields.io/github/downloads/xpack-dev-tools/xbb-bootstrap-xpack/total.svg)](https://github.com/xpack-dev-tools/xbb-bootstrap-xpack/releases/)
  - [individual file counters](https://somsubhra.github.io/github-release-stats/?username=xpack-dev-tools&repository=xbb-bootstrap-xpack) (grouped per release)
- npmjs.com [`@xpack-dev-tools/xbb-bootstrap`](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap/) xPack
  - latest release, per month
[![npm (scoped)](https://img.shields.io/npm/v/@xpack-dev-tools/xbb-bootstrap.svg)](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap/)
[![npm](https://img.shields.io/npm/dm/@xpack-dev-tools/xbb-bootstrap.svg)](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap/)
  - all releases [![npm](https://img.shields.io/npm/dt/@xpack-dev-tools/xbb-bootstrap.svg)](https://www.npmjs.com/package/@xpack-dev-tools/xbb-bootstrap/)

Credit to [Shields IO](https://shields.io) for the badges and to
[Somsubhra/github-release-stats](https://github.com/Somsubhra/github-release-stats)
for the individual file counters.


CHANGELOG.md
# Change & release log

Entries in this file are in reverse chronological order.

## 2022-10-03

* v4.0.0 published on npmjs.com
* v4.0 released

## 2022-10-01

* v4.0 prepared

## 2022-09-13

* created via copy/paste from OpenOCD


scripts/tests/common-functions-source.sh
# -----------------------------------------------------------------------------
# This file is part of the xPack distribution.
#   (https://xpack.github.io)
# Copyright (c) 2020 Liviu Ionescu.
#
# Permission to use, copy, modify, and/or distribute this software
# for any purpose is hereby granted, under the terms of the MIT license.
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# Common functions used in various tests.
#
# Requires
# - app_folder_path
# - test_folder_path
# - archive_platform (win32|linux|darwin)

# -----------------------------------------------------------------------------

# TEST_BIN_PATH - folder with the tested binaries
# TESTS_FOLDER_PATH - temporary folder where individual tests are performed

function run_tests()
{
  echo
  env | sort

  # Removed, they use absolute paths.
  # test_autoconf "${TEST_BIN_PATH}"
  # test_autogen "${TEST_BIN_PATH}"
  # test_automake "${TEST_BIN_PATH}"

  test_bash "${TEST_BIN_PATH}"
  test_bison "${TEST_BIN_PATH}"
  test_coreutils "${TEST_BIN_PATH}"
  test_curl "${TEST_BIN_PATH}"
  test_diffutils "${TEST_BIN_PATH}"
  test_dos2unix "${TEST_BIN_PATH}"

  if false # [ "${TARGET_PLATFORM}" == "darwin" ] && [ "${TARGET_ARCH}" == "arm64" ]
  then
    :
  else
    : # test_flex "${TEST_BIN_PATH}"
  fi

  test_gawk "${TEST_BIN_PATH}"
  test_gettext "${TEST_BIN_PATH}"
  test_gnutls "${TEST_BIN_PATH}"
  test_gpg "${TEST_BIN_PATH}"

  # Removed, on Linux it fails with --version
  # test_guile "${TEST_BIN_PATH}"

  test_m4 "${TEST_BIN_PATH}"
  test_make "${TEST_BIN_PATH}"
  test_makedepend "${TEST_BIN_PATH}"
  test_p7zip "${TEST_BIN_PATH}"
  test_patch "${TEST_BIN_PATH}"
  test_patchelf "${TEST_BIN_PATH}"
  test_perl "${TEST_BIN_PATH}"
  test_pkg_config "${TEST_BIN_PATH}"
  test_python3 "${TEST_BIN_PATH}"

  if [ "${TARGET_PLATFORM}" == "darwin" ]
  then
    test_realpath "${TEST_BIN_PATH}"
  fi

  test_rhash "${TEST_BIN_PATH}"
  test_re2c "${TEST_BIN_PATH}"
  test_sed "${TEST_BIN_PATH}"
  test_tar "${TEST_BIN_PATH}"
  test_tcl "${TEST_BIN_PATH}"
  test_texinfo "${TEST_BIN_PATH}"
  test_wget "${TEST_BIN_PATH}"
}

function update_image()
{
  local image_name="$1"

  # Make sure that the minimum prerequisites are met.
  if [[ ${image_name} == github-actions-ubuntu* ]]
  then
    : # sudo apt-get -qq install -y XXX
  elif [[ ${image_name} == *ubuntu* ]] || [[ ${image_name} == *debian* ]] || [[ ${image_name} == *raspbian* ]]
  then
    run_verbose apt-get -qq update
    run_verbose apt-get -qq install -y git-core curl tar gzip lsb-release binutils
    run_verbose apt-get -qq install -y g++ libc6-dev libstdc++6
  elif [[ ${image_name} == *centos* ]] || [[ ${image_name} == *redhat* ]] || [[ ${image_name} == *fedora* ]]
  then
    run_verbose yum install -y -q git curl tar gzip redhat-lsb-core binutils
    run_verbose yum install -y -q diffutils
    run_verbose yum install -y -q gcc-c++ glibc-devel libstdc++-devel
  elif [[ ${image_name} == *suse* ]]
  then
    run_verbose zypper -q --no-gpg-checks in -y git-core curl tar gzip lsb-release binutils findutils util-linux
    run_verbose zypper -q --no-gpg-checks in -y diffutils
    run_verbose zypper -q --no-gpg-checks in -y gcc-c++ glibc-devel libstdc++6
  elif [[ ${image_name} == *manjaro* ]]
  then
    # run_verbose pacman-mirrors -g
    run_verbose pacman -S -y -q --noconfirm

    # Update even if up to date (-yy) & upgrade (-u).
    # pacman -S -yy -u -q --noconfirm
    run_verbose pacman -S -q --noconfirm --noprogressbar git curl tar gzip lsb-release binutils
    run_verbose pacman -S -q --noconfirm --noprogressbar diffutils # For cmp
    run_verbose pacman -S -q --noconfirm --noprogressbar gcc gcc-libs
  elif [[ ${image_name} == *archlinux* ]]
  then
    run_verbose pacman -S -y -q --noconfirm

    # Update even if up to date (-yy) & upgrade (-u).
    # pacman -S -yy -u -q --noconfirm
    run_verbose pacman -S -q --noconfirm --noprogressbar git curl tar gzip lsb-release binutils
    run_verbose pacman -S -q --noconfirm --noprogressbar diffutils # For cmp
    run_verbose pacman -S -q --noconfirm --noprogressbar gcc gcc-libs
  fi
}

# -----------------------------------------------------------------------------


scripts/defs-source.sh
# -----------------------------------------------------------------------------
# This file is part of the xPack distribution.
#   (https://xpack.github.io)
# Copyright (c) 2019 Liviu Ionescu.
#
# Permission to use, copy, modify, and/or distribute this software
# for any purpose is hereby granted, under the terms of the MIT license.
# -----------------------------------------------------------------------------

# Application definitions used in the build scripts.
# As the name implies, it should contain only definitions and should
# be included with 'source' by the host and container scripts.

# Warning: MUST NOT depend on $HOME or other environment variables.

# -----------------------------------------------------------------------------

# Used to display the application name.
APP_NAME=${APP_NAME:-"XBB Bootstrap"}

# Used as part of file/folder paths.
APP_LC_NAME=${APP_LC_NAME:-"xbb-bootstrap"}

DISTRO_NAME=${DISTRO_NAME:-"xPack"}
DISTRO_LC_NAME=${DISTRO_LC_NAME:-"xpack"}
DISTRO_TOP_FOLDER=${DISTRO_TOP_FOLDER:-"xPacks"}

APP_DESCRIPTION="${DISTRO_NAME} ${APP_NAME}"

# -----------------------------------------------------------------------------

GITHUB_ORG="${GITHUB_ORG:-"xpack-dev-tools"}"
GITHUB_REPO="${GITHUB_REPO:-"${APP_LC_NAME}-xpack"}"
GITHUB_PRE_RELEASES="${GITHUB_PRE_RELEASES:-"pre-releases"}"

NPM_PACKAGE="${NPM_PACKAGE:-"@xpack-dev-tools/${APP_LC_NAME}@next"}"

# -----------------------------------------------------------------------------


scripts/common-versions-source.sh
# -----------------------------------------------------------------------------
# This file is part of the xPacks distribution.
#   (https://xpack.github.io)
# Copyright (c) 2019 Liviu Ionescu.
#
# Permission to use, copy, modify, and/or distribute this software
# for any purpose is hereby granted, under the terms of the MIT license.
# -----------------------------------------------------------------------------

# Helper script used in the xPack build scripts. As the name implies,
# it should contain only functions and should be included with 'source'
# by the build scripts (both native and container).

# -----------------------------------------------------------------------------

function build_versions()
{
  # Don't use a comma since the regular expression
  # that processes this string in the Makefile, silently fails and the
  # bfdver.h file remains empty.
  # BRANDING="${DISTRO_NAME} ${APP_NAME} ${TARGET_MACHINE}"

  # TEST_PATH="${APP_PREFIX}"
  # TEST_BIN_PATH="${TEST_PATH}/bin"

  # Install all binaries in the public arena.
  BINS_INSTALL_FOLDER_PATH="${APP_INSTALL_FOLDER_PATH}"

  # Keep them in sync with combo archive content.
  if [[ "${RELEASE_VERSION}" =~ 4\.0 ]]
  then
    (
      libtool_version="2.4.6"

      # -----------------------------------------------------------------------
      # All compiled with the Linux/macOS native compiler, no Windows.

      xbb_activate

      if [ "${TARGET_PLATFORM}" == "darwin" ]
      then
        prepare_clang_env
      fi

      # -----------------------------------------------------------------------

      if [ "${TARGET_PLATFORM}" == "darwin" ]
      then
        # Note: Makefile calls gcc explicitly.
        build_realpath "1.0.0"
      fi

      # -----------------------------------------------------------------------

      # New zlib, used in most of the tools.
      # depends=('glibc')
      build_zlib "1.2.11"

      build_bzip2 "1.0.8"

      # Libraries, required by gcc.
      # depends=('gcc-libs' 'sh')
      build_gmp "6.2.1"
      # depends=('gmp>=5.0')
      build_mpfr "4.1.0"
      # depends=('mpfr')
      build_mpc "1.2.1"

      (
        if [ "${TARGET_PLATFORM}" == "darwin" ] && [[ ${CC} =~ .*gcc.* ]]
        then
          # The GCC linker fails with an assert:
          # ld: Assertion failed: (_file->_atomsArrayCount == computedAtomCount && "more atoms allocated than expected"), function parse, file macho_relocatable_file.cpp, line 2061.
          prepare_clang_env
        fi

        # depends=('gmp')
        build_isl "0.24"
      )

      # -----------------------------------------------------------------------

      # Replacement for the old libcrypt.so.1.
      build_libxcrypt "4.4.26"

      (
        if [ "${TARGET_PLATFORM}" == "darwin" ] && [[ ${CC} =~ .*gcc.* ]]
        then
          # The GCC linker fails with an assert:
          # ld: Assertion failed: (_file->_atomsArrayCount == computedAtomCount && "more atoms allocated than expected"), function parse, file macho_relocatable_file.cpp, line 2061.
          prepare_clang_env
        fi

        # depends=('perl')
        build_openssl "1.1.1q" # "1.1.1l"
      )

      # Libraries, required by gnutls.
      # depends=('glibc')
      build_libtasn1 "4.18.0"

      # Library, required by Python.
      # depends=('glibc')
      build_expat "2.4.1"

      # depends=('glibc')
      build_libffi "3.4.2"

      # Library, required by libunistring, wget.
      # depends=()
      # Harmful for GCC 9.
      build_libiconv "1.16"

      build_libunistring "0.9.10"

      # Required by Python
      build_mpdecimal "2.5.1"

      # Libary, required by tar.
      # depends=('sh')
      build_xz "5.2.5"

      # Requires openssl.
      # depends=('glibc' 'gmp')
      # PATCH!
      build_nettle "3.7.3"

      # Required by bash, readline
      NCURSES_DISABLE_WIDEC="y"
      build_ncurses "6.2"

      # depends=('glibc' 'ncurses' 'libncursesw.so')
      build_readline "8.1"

      build_sqlite "3390200"

      build_libxml2 "2.10.2"

      # -----------------------------------------------------------------------
      # Tools

      # depends=('glibc' 'glib2 (internal)')
      build_pkg_config "0.29.2"

      # depends=('ca-certificates' 'krb5' 'libssh2' 'openssl' 'zlib' 'libpsl' 'libnghttp2')
      build_curl "7.80.0"

      # tar with xz support.
      # depends=('glibc')
      build_tar "1.34"

      (
        # Hmmm... Not standalone, it remembers xbb-gcc and other settings.
        set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"
        tests_add set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"

        # Required before guile.
        # TODO
        build_libtool "${libtool_version}"
      )

      # Required by guile.
      build_gc "8.0.6"

      # depends=('glibc' 'glib2' 'libunistring' 'ncurses')
      build_gettext "0.21"

      (
        # Hmmm... It fails to start, it goes to a prompt no metter what.
        set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"
        tests_add set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"

        # depends=(gmp libltdl ncurses texinfo libunistring gc libffi)
        # 3.x is too new, autogen requires 2.x
        build_guile "2.2.7"
      )

      # depends=('readline>=7.0' glibc ncurses)
      # "5.1" fails on amd64 with:
      # bash-5.1/bashline.c:65:10: fatal error: builtins/builtext.h: No such file or directory
      build_bash "5.1.8"

      (
        # Hmmm... Not standalone, it remembers absolute paths.
        set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"
        tests_add set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"

        # Requires guile 2.x.
        build_autogen "5.18.16"
      )

      # After autogen, requires libopts.so.25.
      # depends=('glibc' 'libidn2' 'libtasn1' 'libunistring' 'nettle' 'p11-kit' 'readline' 'zlib')
      build_gnutls "3.7.2"

      # -----------------------------------------------------------------------
      # GNU tools

      # "8.32" fails on aarch64 with:
      # coreutils-8.32/src/ls.c:3026:24: error: 'SYS_getdents' undeclared (first use in this function); did you mean 'SYS_getdents64'?
      build_coreutils "9.0"

      # depends=('glibc')
      # PATCH!
      # "1.4.19" tests fail on amd64.
      build_m4 "1.4.19"

      # depends=('glibc' 'mpfr')
      build_gawk "5.1.1"

      # depends ?
      build_sed "4.8"

      (
        # Hmmm... Not standalone, it remembers absolute paths.
        set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"
        tests_add set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"

        # depends=('sh' 'perl' 'awk' 'm4' 'texinfo')
        build_autoconf "2.71"
        # depends=('sh' 'perl')

        # PATCH!
        build_automake "1.16.5"
      )

      # depends=('glibc' 'attr')
      build_patch "2.7.6"

      # depends=('libsigsegv')
      build_diffutils "3.8"

      # depends=('glibc')
      build_bison "3.8.2"

      # depends=('glibc' 'guile')
      # PATCH!
      build_make "4.3"

      # -----------------------------------------------------------------------
      # Third party tools

      # depends=('libutil-linux' 'gnutls' 'libidn' 'libpsl>=0.7.1-3' 'gpgme')
      # "1.21.[12]" fails on macOS with
      # lib/malloc/dynarray-skeleton.c:195:13: error: expected identifier or '(' before numeric constant
      # 195 | __nonnull ((1))
      build_wget "1.20.3"

      # Required to build PDF manuals.
      # depends=('coreutils')
      build_texinfo "6.8"

      # depends ?
      # Warning: buggy!
      # "0.12" weird tag
      build_patchelf "0.14.3"

      # depends=('glibc')
      build_dos2unix "7.4.2"

      # Most probably not needed.
      # macOS 10.13: 2.5.35
      # macOS 11.1: 2.6.4
      # Ubuntu 18: 2.6.4
      if false # [ "${TARGET_PLATFORM}" == "darwin" ] && [ "${TARGET_ARCH}" == "arm64" ]
      then
        : # Still problematic, fails to run
      else
        build_flex "2.6.4"
      fi

      # macOS 10.1[03] uses 5.18.2.
      # macOS 11.6 uses 5.30.2
      # HiRes.c:2037:17: error: use of undeclared identifier 'CLOCK_REALTIME'
      #     clock_id = CLOCK_REALTIME;
      #
      # depends=('gdbm' 'db' 'glibc')
      # old PATCH!
      build_perl "5.34.0"

      # Give other a chance to use it.
      # However some (like Python) test for Tk too.
      build_tcl "8.6.12"

      # depends=('curl' 'libarchive' 'shared-mime-info' 'jsoncpp' 'rhash')
      # Already a binary xPack
      # build_cmake "3.22.1"

      # Requires scons
      # depends=('python2')
      # Already a binary xPack
      # build_ninja "1.10.2"

      # depends=('curl' 'expat>=2.0' 'perl-error' 'perl>=5.14.0' 'openssl' 'pcre2' 'grep' 'shadow')
      # Ubuntu 18: 2.17.1
      # macOS 10.13: 2.17.2
      # macOS 11.6: 2.30.1
      # Hopefully no longer needed, it is very heavy,
      # libexec/git-core takes 475 MB.
      # build_git "2.34.1"

      # 17.04 is from a fork easier to build on macOS.
      build_p7zip "17.04" # "16.02"

      # "1.4.[12]" fail on amd64 with
      # librhash/librhash.so.0: undefined reference to `aligned_alloc'
      # For Apple Silicon, use 1.4.3 or higher.
      build_rhash "1.4.3" # "1.4.2"

      build_re2c "2.2"

      # -----------------------------------------------------------------------

      # $1=nvm_version
      # $2=node_version
      # $3=npm_version
      # build_nvm "0.35.2"

      # -----------------------------------------------------------------------

      (
        set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"
        tests_add set_bins_install "${LIBS_INSTALL_FOLDER_PATH}"

        build_libgpg_error "1.42"
        # "1.9.3" Fails many tests on macOS
        build_libgcrypt "1.9.4"
        build_libassuan "2.5.5"
        # patched
        build_libksba "1.6.0"

        build_npth "1.6"
      )

      # "2.3.1" fails on macOS 10.13, requires libgcrypt 1.9
      # "2.2.28" fails on amd64
      build_gnupg  "2.3.3"

      # -----------------------------------------------------------------------

      # makedepend is needed by openssl
      build_xorg_util_macros "1.19.3"
      # PATCH!
      build_xorg_xproto "7.0.31" # Needs a patch for aarch64.
      build_makedepend "1.0.6"

      # -----------------------------------------------------------------------

      # Avoid Java for now, not longer available on early Apple Silicon.
      if false
      then
        build_ant "1.10.12" # "1.10.10" # "1.10.7"
        build_maven "3.8.3" # "3.8.1" # "3.6.3"
      fi

      # -----------------------------------------------------------------------

      # mingw-w64 & wine migrated to separate packages.

      # -----------------------------------------------------------------------
      # Python family.
      if true
      then
        # macOS 10.13/11.6 use 2.7.16, close enough.
        # On Apple Silicon it fails, it is not worth the effort.
        # On Ubuntu 18 there is 2.7.17; not much difference with 2.7.18.
        # depends=('bzip2' 'gdbm' 'openssl' 'zlib' 'expat' 'sqlite' 'libffi')
        # build_python2 "2.7.18"

        # homebrew: gdbm, mpdecimal, openssl, readline, sqlite, xz; bzip2, expat, libffi, ncurses, unzip, zlib
        # arch: 'bzip2' 'expat' 'gdbm' 'libffi' 'libnsl' 'libxcrypt' 'openssl' 'zlib'
        build_python3 "3.9.9"

        # The necessary bits to build these optional modules were not found:
        # _bz2                  _dbm                  _gdbm
        # _sqlite3              _tkinter              _uuid
        # Failed to build these modules:
        # _curses               _curses_panel         _decimal

        # depends=('python3')
        # "4.1.0" fails on macOS 10.13
        # TODO
        # build_scons "4.2.0"

        # TODO
        # build_sphinx "4.3.0"

        # skipped, already available as an xPack
        # build_meson "0.60.2"
      fi

      echo
      echo "Removing manuals, docs, info, html..."
      rm -rfv "${APP_PREFIX}/man" # bzip2
      rm -rfv "${APP_PREFIX}/share/doc"
      rm -rfv "${APP_PREFIX}/share/man"
      rm -rfv "${APP_PREFIX}/share/info"
      rm -rfv "${APP_PREFIX}/share/gtk-doc/html"

      (
        echo
        echo "Patching perl scripts..."

        set +e
        cd "${APP_PREFIX}/bin"

        for f in $(grep "#!${APP_PREFIX}/bin/perl" * | sed -e "s|:.*||")
        do
          run_verbose sed -i -e "s|#!${APP_PREFIX}/bin/perl|#! perl|" $f
        done

        for f in $(grep "eval 'exec ${APP_PREFIX}/bin/perl" * | sed -e "s|:.*||")
        do
          run_verbose sed -i -e "s|exec ${APP_PREFIX}/bin/perl|exec perl|" $f
        done

        for f in $(grep "#! /opt/xbb/bin/perl" * | sed -e "s|:.*||")
        do
          run_verbose sed -i -e "s|#! /opt/xbb/bin/perl|#! perl|" $f
        done

        for f in $(grep "#! ${HOME}/.local/xbb/bin/perl" * | sed -e "s|:.*||")
        do
          echo $f
          run_verbose sed -i -e "s|#! ${HOME}/.local/xbb/bin/perl|#! perl|" $f
        done
      )

      (
        echo
        echo "Patching python scripts..."

        set +e
        cd "${APP_PREFIX}/bin"

        for f in $(grep "#!${APP_PREFIX}/bin/python3" * | sed -e "s|:.*||")
        do
          run_verbose sed -i -e "s|#!${APP_PREFIX}/bin/python3.*|#! python3|" $f
        done
      )
      (
        cd "${APP_PREFIX}/bin"
        echo
        echo "List remaining xbb references in /bin..."
        grep -i 'xbb' * | grep -v 'Binary file'
      )
    )
    # -------------------------------------------------------------------------
  else
    echo "Unsupported version ${RELEASE_VERSION}."
    exit 1
  fi
}

# -----------------------------------------------------------------------------
